pkg/registry/registry.go

```diff
@@ -54,9 +60,122 @@ func (o RegistryClient) DownloadFileInMemory(params dfutil.HTTPRequestParams) ([
+	// Let the project be downloaded in a temp directory
+	starterProjectTmpDir, err := o.fsys.TempDir("", "odostarterproject")
+	if err != nil {
+		return err
+	}
+	err = o.fsys.MkdirAll(starterProjectTmpDir, 0755)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		ferr := o.fsys.RemoveAll(starterProjectTmpDir)
+		if ferr != nil {
+			klog.V(2).Infof("failed to delete temporary starter project dir %s", starterProjectTmpDir)
+		}
+	}()
+
+	downloadStarterProjectErr := DownloadStarterProject(o.fsys, starterProject, decryptedToken, starterProjectTmpDir, verbose)
+	if downloadStarterProjectErr != nil {
+		return downloadStarterProjectErr
+	}
+
+	// Case 1
+	if containsDevfile, _ := location.DirectoryContainsDevfile(o.fsys, starterProjectTmpDir); containsDevfile {
+		fmt.Println()
+		log.Warning("devfile.yaml is present inside the starter project; replacing the entire content of the current directory with the starter project")
+		err = removeDirectoryContents(contextDir, o.fsys)
+		if err != nil {
+			klog.V(0).Infof(err.Error())
+		}
+		return util.CopyDirWithFS(starterProjectTmpDir, contextDir, o.fsys)
+	}
+
+	// Case 2 & 3
+	conflictingFiles, cerr := isConflicting(starterProjectTmpDir, contextDir, o.fsys)
+	if cerr != nil {
+		return cerr
+	}
+	// Case 3
+	if len(conflictingFiles) == 0 {
+		copyErr := util.CopyDirWithFS(starterProjectTmpDir, contextDir, o.fsys)
+		if copyErr != nil {
+			return copyErr
+		}
+	} else {
+		// Case 2
+		err = o.fsys.MkdirAll(filepath.Join(contextDir, CONFLICT_DIR_NAME), 0750)
+		if err != nil {
+			return err
+		}
+
+		copyErr := util.CopyDirWithFS(starterProjectTmpDir, filepath.Join(contextDir, CONFLICT_DIR_NAME), o.fsys)
+		if copyErr != nil {
+			return copyErr
+		}
+		log.Warningf("\nThere are conflicting files (%s) between starter project and the current directory, hence the starter project has been copied to %s", strings.Join(conflictingFiles, ", "), filepath.Join(contextDir, CONFLICT_DIR_NAME))
+	}
+	return nil
+}
+
+// removeDirectoryContents attempts to remove dir contents without deleting the directory itself, unlike os.RemoveAll() method
+func removeDirectoryContents(path string, fsys filesystem.Filesystem) error {
+	dir, err := fsys.ReadDir(path)
+	if err != nil {
+		return err
+	}
+	for _, f := range dir {
+		// a bit of cheating by using absolute file name to make sure this works with a fake filesystem, especially a memory map which is used by our unit tests
+		// memorymap's Name() method trims the full path and returns just the file name, which then becomes impossible to find by the RemoveAll method that looks for prefix
+		// See: https://github.com/redhat-developer/odo/blob/d717421494f746a5cb12da135f561d12750935f3/vendor/github.com/spf13/afero/memmap.go#L282
+		absFileName := filepath.Join(path, f.Name())
+		err = fsys.RemoveAll(absFileName)
+		if err != nil {
+			return fmt.Errorf("failed to remove %s; cause: %s", absFileName, err)
```

Comment:
In case we need to `Unwrap` it.

```suggestion
			return fmt.Errorf("failed to remove %s; cause: %w", absFileName, err)
```